compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolResolver.java

```diff
@@ -1345,14 +1345,28 @@ public void visit(BLangFiniteTypeNode finiteTypeNode) {
 
         BFiniteType finiteType = new BFiniteType(finiteTypeSymbol);
         for (BLangExpression literal : finiteTypeNode.valueSpace) {
-            ((BLangLiteral) literal).setBType(symTable.getTypeFromTag(((BLangLiteral) literal).getBType().tag));
+            literal.setBType(getTypeOfNumericLiteral(literal.getKind(), literal.getBType()));
             finiteType.addValue(literal);
         }
         finiteTypeSymbol.type = finiteType;
 
         resultType = finiteType;
     }
 
+public BType getTypeOfNumericLiteral(NodeKind nodeKind, BType type) {
+    if (nodeKind != NodeKind.NUMERIC_LITERAL || type.getKind() != TypeKind.UNION) {
+        return type;
+    }
+    Set<BType> memberTypes = ((BUnionType) type).getMemberTypes();
```

Comment:
Can you share examples for when these are hit?

Reply:
For these kinds of scenarios
```ballerina
public function main() {
    "value" _ = "value";
}

```